Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,826 @@
+            m           -> throw $ GMEString $ "ImportedFrom: could not qualify " ++ name ++ " from these exports: " ++ show qualifiedNames ++ "\n    matches: " ++ show m
+
+        -- Time for some stringly typed rubbish. The previous test used
+        -- postfixMatch but this failed on an import that had "hiding (lines, unlines)" since
+        -- both lines and unlines matched. Prepending a dot doesn't work due to things like ".=" from
+        -- Control.Lens. So we manually check that the suffix matches, that the next symbol is a dot,
+        -- and then an alpha character, which hopefully is the end of a module name. Such a mess.
+        where f n qn = if length qn - length n - 2 >= 0
+                            then n `isSuffixOf` qn && isAlpha (qn !! (length qn - length n - 2)) && (qn !! (length qn - length n - 1)) == '.'
+                            else throw $ GMEString $ "ImportedFrom internal error: trying to check if \"" ++ n ++ "\" is a match for \"" ++ qn ++ "\""
+
+refineExportsIt :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineExportsIt mysymbol exports = map (\e -> e { qualifiedExports = f symbol e }) exports
+  where
+    -- Deal with these?
+    symbol = case mysymbol of
+                MySymbolSysQualified  s -> s
+                MySymbolUserQualified s -> s
+
+    f sym export = filter (postfixMatch sym) thisExports
+       where thisExports = qualifiedExports export         -- Things that this module exports.
+
+-- On an internal symbol (e.g. Show), refineExportsIt erronously filters out everything.
+-- For example mnsymbol = "base-4.9.0.0:GHC.Show.C:Show" and the matching
+-- name "base-4.9.0.0:GHC.Show.Show" from the Prelude. The problem seems to be the
+-- module name GHC.Show.C, probably referring to an internal C library.
+--
+-- To get around this, refineExportsItFallbackInternal uses a less strict matching
+-- rule. If the 'stage3' results are empty we fall back to this refiner.
+refineExportsItFallbackInternal :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineExportsItFallbackInternal mysymbol exports
+    = case separateBy ':' symbol of
+        [p, _, x]   -> map (\e -> e { qualifiedExports = f p x e }) exports
+        _           -> exports
+  where
+    -- Deal with these?
+    symbol = case mysymbol of
+                MySymbolSysQualified  s -> s
+                MySymbolUserQualified s -> s
+
+    -- Check if package name matches and postfix symbol matches (roughly).
+    f p sym export = filter
+                        (\z -> p `isPrefixOf` z && postfixMatch sym z)
+                        (qualifiedExports export)
+
+refineLeadingDot :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineLeadingDot (MySymbolUserQualified _)           exports = exports
+refineLeadingDot (MySymbolSysQualified symb)         exports = map (\e -> e { qualifiedExports = f leadingDot e }) exports
+  where
+    leadingDot :: String
+    leadingDot = '.' : last (separateBy '.' symb)
```

Comment:
Technically this is ok. At some point I want to refactor the types so that these operations aren't so stringy.

```
commit c995e15277530ad43a5b4aa73e0caab02c89f6f5
Author: Carlo Hamalainen <carlo@carlo-hamalainen.net>
Date:   Tue Aug 9 13:05:30 2016 +0800

    Note on use of ```last```.

diff --git a/Language/Haskell/GhcMod/ImportedFrom.hs b/Language/Haskell/GhcMod/ImportedFrom.hs
index 3ee0170..abd230e 100644
--- a/Language/Haskell/GhcMod/ImportedFrom.hs
+++ b/Language/Haskell/GhcMod/ImportedFrom.hs
@@ -535,6 +535,9 @@ refineLeadingDot :: MySymbol -> [ModuleExports] -> [ModuleExports]
 refineLeadingDot (MySymbolUserQualified _)           exports = exports
 refineLeadingDot (MySymbolSysQualified symb)         exports = map (\e -> e { qualifiedExports = f leadingDot e }) exports
   where
+    -- We use leadingDot only when we have an 'MySymbolSysQualified symb' so
+    -- the 'last' will be ok. Sample value of 'symb' in this case is
+    -- "base-4.8.2.0:Data.Foldable.length".
     leadingDot :: String
     leadingDot = '.' : last (splitOn "." symb)
```